File: go/test/alert_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/cheapshark-sdk"
	"github.com/voxgig-sdk/cheapshark-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestAlertEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Alert(nil)
		if ent == nil {
			t.Fatal("expected non-nil AlertEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := alertBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "alert." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set CHEAPSHARK_TEST_ALERT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		alertRef01Ent := client.Alert(nil)
		alertRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "alert"}, setup.data), "alert_ref01"))

		alertRef01DataResult, err := alertRef01Ent.Create(alertRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		alertRef01Data = core.ToMapAny(alertRef01DataResult)
		if alertRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LIST
		alertRef01Match := map[string]any{}

		alertRef01ListResult, err := alertRef01Ent.List(alertRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		alertRef01List, alertRef01ListOk := alertRef01ListResult.([]any)
		if !alertRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", alertRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(alertRef01List), map[string]any{"id": alertRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// REMOVE
		alertRef01MatchRm0 := map[string]any{
			"id": alertRef01Data["id"],
		}
		_, err = alertRef01Ent.Remove(alertRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		alertRef01MatchRt0 := map[string]any{}

		alertRef01ListRt0Result, err := alertRef01Ent.List(alertRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		alertRef01ListRt0, alertRef01ListRt0Ok := alertRef01ListRt0Result.([]any)
		if !alertRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", alertRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(alertRef01ListRt0), map[string]any{"id": alertRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func alertBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "alert", "AlertTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read alert test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse alert test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"alert01", "alert02", "alert03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("CHEAPSHARK_TEST_ALERT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"CHEAPSHARK_TEST_ALERT_ENTID": idmap,
		"CHEAPSHARK_TEST_LIVE":      "FALSE",
		"CHEAPSHARK_TEST_EXPLAIN":   "FALSE",
		"CHEAPSHARK_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["CHEAPSHARK_TEST_ALERT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["CHEAPSHARK_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["CHEAPSHARK_APIKEY"],
			},
			extra,
		})
		client = sdk.NewCheapsharkSDK(core.ToMapAny(mergedOpts))
	}

	live := env["CHEAPSHARK_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["CHEAPSHARK_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/AlertEntityTest.php
<?php
declare(strict_types=1);

// Alert entity test

require_once __DIR__ . '/../cheapshark_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class AlertEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = CheapsharkSDK::test(null, null);
        $ent = $testsdk->Alert(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "alert" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = CheapsharkSDK::test($seed, null);
        $seen = iterator_to_array($base->Alert(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = CheapsharkConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = CheapsharkSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Alert(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = alert_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "alert." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set CHEAPSHARK_TEST_ALERT_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $alert_ref01_ent = $client->Alert(null);
        $alert_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.alert"), "alert_ref01"));

        $alert_ref01_data_result = $alert_ref01_ent->create($alert_ref01_data, null);
        $alert_ref01_data = Helpers::to_map($alert_ref01_data_result);
        $this->assertNotNull($alert_ref01_data);

        // LIST
        $alert_ref01_match = [];

        $alert_ref01_list_result = $alert_ref01_ent->list($alert_ref01_match, null);
        $this->assertIsArray($alert_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($alert_ref01_list_result),
            ["id" => $alert_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // REMOVE
        $alert_ref01_match_rm0 = [
            "id" => $alert_ref01_data["id"],
        ];
        $alert_ref01_ent->remove($alert_ref01_match_rm0, null);

        // LIST
        $alert_ref01_match_rt0 = [];

        $alert_ref01_list_rt0_result = $alert_ref01_ent->list($alert_ref01_match_rt0, null);
        $this->assertIsArray($alert_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($alert_ref01_list_rt0_result),
            ["id" => $alert_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function alert_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/alert/AlertTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = CheapsharkSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["alert01", "alert02", "alert03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("CHEAPSHARK_TEST_ALERT_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "CHEAPSHARK_TEST_ALERT_ENTID" => $idmap,
        "CHEAPSHARK_TEST_LIVE" => "FALSE",
        "CHEAPSHARK_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["CHEAPSHARK_TEST_ALERT_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["CHEAPSHARK_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new CheapsharkSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["CHEAPSHARK_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["CHEAPSHARK_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
